File: packages/go/sdk/codec.go
// Package sdk is the GoNext Go plugin SDK, TinyGo-targeted.
//
// A plugin author writes:
//
//	package main
//
//	import "github.com/Singleton-Solution/GoNext/packages/go/sdk"
//
//	func main() {
//	    sdk.RegisterAction("posts.publish", func(args []any) error {
//	        sdk.Host.KV.Set("last-published", []byte(args[0].(string)))
//	        sdk.Host.Audit.Emit("post.indexed", map[string]any{"id": args[0]})
//	        return nil
//	    })
//	    sdk.RegisterFilter("the_content", func(value []byte, args []any) ([]byte, error) {
//	        return append(value, []byte("\n<!-- enhanced -->")...), nil
//	    })
//	    sdk.PluginInit(sdk.NewManifest("hello", "0.1.0").
//	        WithCapability("kv.write").
//	        WithCapability("audit.emit").
//	        WithAction("posts.publish").
//	        WithFilter("the_content").
//	        Build())
//	}
//
// `tinygo build -target=wasi -o plugin.wasm .` produces a module that
// satisfies the v1 hook ABI in packages/go/plugins/abi/hooks — same
// gn_handle_hook / gn_alloc / gn_free / gn_panic exports the host
// expects.
//
// # Wire format
//
// The SDK encodes/decodes payloads on the same JSON envelope shape the
// host uses in packages/go/plugins/abi/hooks/marshal.go. We mirror the
// shapes here rather than depending on the host package so the SDK has
// no transitive imports the TinyGo toolchain can't handle.
//
// # Build modes
//
// codec.go (this file) is pure Go and compiles with both the stock
// toolchain AND TinyGo. Codec unit tests run on the stock toolchain so
// CI can validate marshalling without requiring TinyGo on PATH.
//
// hooks.go, host.go, host_*.go, and runtime_wasm.go contain the
// guest-side runtime that only makes sense inside a wasm32-wasi build.
// They use build tags to switch between a "real" implementation (under
// the wasi tag) and a no-op stub (everywhere else) — so a plugin
// author writing tests against their own handler functions can build
// the package under the stock toolchain too.
package sdk

import (
	"encoding/json"
	"errors"
	"fmt"
)

// PayloadKind tags whether a payload carries action arguments (one
// list) or filter arguments (a value + extras list). Mirrors
// abi/hooks.PayloadKind byte-for-byte; the constant strings are part
// of the v1 wire format.
type PayloadKind string

const (
	// PayloadKindAction is the action-call shape: a list of args.
	PayloadKindAction PayloadKind = "action"

	// PayloadKindFilter is the filter-call shape: a transformable
	// value plus the per-call extras.
	PayloadKindFilter PayloadKind = "filter"
)

// ActionPayload mirrors abi/hooks.ActionPayload — the wire form of an
// action-call payload. Kept here so the SDK has no host-package
// imports (TinyGo can't compile the host module).
type ActionPayload struct {
	Kind PayloadKind   `json:"kind"`
	Args []interface{} `json:"args"`
}

// FilterPayload mirrors abi/hooks.FilterPayload.
type FilterPayload struct {
	Kind  PayloadKind     `json:"kind"`
	Value json.RawMessage `json:"value"`
	Args  []interface{}   `json:"args"`
}

// FilterResult mirrors abi/hooks.FilterResult — the envelope a filter
// handler returns. The host decodes this shape on the way out.
type FilterResult struct {
	Value json.RawMessage `json:"value"`
}

// ResultStatus mirrors abi/hooks.ResultStatus. Sentinels are negative
// int32s the SDK packs into the low half of the i64 result when no
// body is being returned.
type ResultStatus int32

const (
	// StatusOK signals success-with-no-body. Returned by every action
	// handler that completes without producing a result envelope.
	StatusOK ResultStatus = 0

	// StatusError is the generic guest-reported failure. The SDK
	// returns this when a handler returns a non-nil error that isn't
	// one of the more specific sentinels below.
	StatusError ResultStatus = -1

	// StatusOutOfMemory signals an allocator failure inside the
	// handler. The SDK returns this when gn_alloc would fail.
	StatusOutOfMemory ResultStatus = -2

	// StatusBadPayload signals the SDK couldn't decode the inbound
	// payload — a host/guest ABI mismatch.
	StatusBadPayload ResultStatus = -3

	// StatusUnknownHook signals no handler is registered for the
	// requested name. The host typically treats this as a stale
	// plugin binary.
	StatusUnknownHook ResultStatus = -4
)

// PackResult composes the i64 the SDK returns from gn_handle_hook
// from a (ptr, len) pair. Pointer occupies the high 32 bits; length
// occupies the low 32 bits as a signed int32 so negative sentinels
// survive the round trip. Mirrors hooks.packResult.
func PackResult(ptr uint32, length int32) uint64 {
	return uint64(ptr)<<32 | uint64(uint32(length))
}

// UnpackResult is the inverse of PackResult.
func UnpackResult(packed uint64) (ptr uint32, length int32) {
	ptr = uint32(packed >> 32)
	length = int32(packed & 0xFFFFFFFF)
	return ptr, length
}

// MarshalActionPayload returns the JSON wire bytes for an action call.
// Mirrors hooks.MarshalActionPayload.
func MarshalActionPayload(args []interface{}) ([]byte, error) {
	if args == nil {
		args = []interface{}{}
	}
	return json.Marshal(ActionPayload{Kind: PayloadKindAction, Args: args})
}

// UnmarshalActionPayload decodes an action-call envelope. Returns an
// error wrapped around ErrBadPayload on a shape mismatch so handlers
// can errors.Is against the sentinel without type-asserting.
func UnmarshalActionPayload(buf []byte) (*ActionPayload, error) {
	if len(buf) == 0 {
		// The bridge always emits a complete envelope, even for
		// zero-arg actions ({"kind":"action","args":[]}). An empty
		// buffer is a contract violation; treat as bad payload.
		return nil, fmt.Errorf("%w: empty payload", ErrBadPayload)
	}
	var p ActionPayload
	if err := json.Unmarshal(buf, &p); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrBadPayload, err)
	}
	if p.Kind != PayloadKindAction {
		return nil, fmt.Errorf("%w: expected %q kind, got %q", ErrBadPayload, PayloadKindAction, p.Kind)
	}
	if p.Args == nil {
		p.Args = []interface{}{}
	}
	return &p, nil
}

// MarshalFilterPayload returns the JSON wire bytes for a filter call.
// value MAY be nil — in that case the envelope carries "value":null.
// Mirrors hooks.MarshalFilterPayload.
func MarshalFilterPayload(value json.RawMessage, args []interface{}) ([]byte, error) {
	if args == nil {
		args = []interface{}{}
	}
	if value == nil {
		value = json.RawMessage("null")
	}
	return json.Marshal(FilterPayload{Kind: PayloadKindFilter, Value: value, Args: args})
}

// UnmarshalFilterPayload decodes a filter-call envelope. Returns an
// error wrapped around ErrBadPayload on a shape mismatch.
func UnmarshalFilterPayload(buf []byte) (*FilterPayload, error) {
	if len(buf) == 0 {
		return nil, fmt.Errorf("%w: empty payload", ErrBadPayload)
	}
	var p FilterPayload
	if err := json.Unmarshal(buf, &p); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrBadPayload, err)
	}
	if p.Kind != PayloadKindFilter {
		return nil, fmt.Errorf("%w: expected %q kind, got %q", ErrBadPayload, PayloadKindFilter, p.Kind)
	}
	if p.Args == nil {
		p.Args = []interface{}{}
	}
	if len(p.Value) == 0 {
		p.Value = json.RawMessage("null")
	}
	return &p, nil
}

// MarshalFilterResult produces the wire bytes the host expects from a
// filter handler return — a JSON object with a single "value" key.
func MarshalFilterResult(value json.RawMessage) ([]byte, error) {
	if value == nil {
		value = json.RawMessage("null")
	}
	return json.Marshal(FilterResult{Value: value})
}

// UnmarshalFilterResult is the inverse: extracts the value bytes from
// a FilterResult envelope.
func UnmarshalFilterResult(buf []byte) (json.RawMessage, error) {
	if len(buf) == 0 {
		return nil, nil
	}
	var r FilterResult
	if err := json.Unmarshal(buf, &r); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrBadPayload, err)
	}
	if len(r.Value) == 0 {
		return json.RawMessage("null"), nil
	}
	return r.Value, nil
}

// Sentinel errors callers can errors.Is against. The SDK never wraps
// arbitrary host errors — the failure modes are categorical, and the
// status code carried in the i64 return value is the canonical signal.
//
// Plugin authors that return errors from their handlers get them
// projected onto one of these statuses by the dispatcher in hooks.go.
var (
	// ErrBadPayload is returned from the Unmarshal* helpers when the
	// inbound envelope can't be decoded as the expected shape.
	ErrBadPayload = errors.New("sdk: bad payload")

	// ErrUnknownHook is returned by the dispatcher when no handler
	// matches the inbound name.
	ErrUnknownHook = errors.New("sdk: unknown hook")

	// ErrOutOfMemory is returned when the SDK's allocator runs out of
	// room. The host translates this to ResultStatusOutOfMemory.
	ErrOutOfMemory = errors.New("sdk: out of memory")

	// ErrHostFailure is returned by the typed Host.* wrappers when a
	// host ABI call returns a negative status (denied, blocked, rate-
	// limited, etc). Inspect the wrapping HostError for the specific
	// status.
	ErrHostFailure = errors.New("sdk: host ABI returned failure status")
)

// HostError wraps a host-ABI failure. Status carries the negative
// sentinel the host returned; Function names which host export was
// called. Implements errors.Is against ErrHostFailure.
type HostError struct {
	Function string
	Status   int32
}

// Error returns a one-line summary.
func (e *HostError) Error() string {
	return fmt.Sprintf("sdk: host call %q failed (status=%d)", e.Function, e.Status)
}

// Is supports errors.Is(err, ErrHostFailure) for typed matching.
func (e *HostError) Is(target error) bool { return target == ErrHostFailure }

File: packages/ts/sdk-plugin/src/codec.ts
/**
 * JSON envelope codec for the GoNext plugin ABI.
 *
 * Mirrors `packages/go/plugins/abi/hooks/marshal.go`. Every payload that
 * crosses the host<->guest boundary is JSON, so the codec is essentially
 * a typed view onto `JSON.stringify` / `JSON.parse` with a few invariants:
 *
 *   - Action payloads always carry an empty array for `args` (never null),
 *     so the guest decoder has a single shape to handle.
 *   - Filter payloads always carry an explicit `value` (JSON-null when
 *     none was supplied) and an empty `args` array when no extras were
 *     passed.
 *   - Filter results carry one field — `value` — also JSON-null when the
 *     handler chose not to transform.
 *
 * Result statuses mirror the negative-int32 sentinels the host returns
 * in the low 32 bits of the packed i64. The Javy guest never returns
 * these directly from JS; the runtime entry point at `src/host.ts`
 * translates a typed throw into the right packed return.
 *
 * Why a dedicated module: the codec is the only piece of the SDK that
 * needs to round-trip cleanly in a Node test environment (Vitest).
 * Keeping it free of `globalThis.gn_*` references means the unit tests
 * for envelope handling don't have to stand up the Javy host shims.
 */

/** Sentinel statuses returned by the host. */
export const ResultStatus = {
  OK: 0,
  Error: -1,
  OutOfMemory: -2,
  BadPayload: -3,
  UnknownHook: -4,
  Trap: -5,
} as const;

/** Type-level enumeration of the result-status sentinels. */
export type ResultStatusCode =
  (typeof ResultStatus)[keyof typeof ResultStatus];

/** Tag for the two payload kinds the hook bus dispatches. */
export const PayloadKind = {
  Action: 'action',
  Filter: 'filter',
} as const;

/** Type-level enumeration of payload kinds. */
export type PayloadKindLiteral =
  (typeof PayloadKind)[keyof typeof PayloadKind];

/**
 * Wire shape of an action payload. Arrays of `unknown` because the bus
 * is variadic over `any` — the guest is responsible for narrowing on a
 * per-hook basis.
 */
export interface ActionPayload {
  kind: typeof PayloadKind.Action;
  args: unknown[];
}

/**
 * Wire shape of a filter payload. `value` is the transformable thing
 * threaded through the chain; `args` carries the per-call extras.
 */
export interface FilterPayload {
  kind: typeof PayloadKind.Filter;
  value: unknown;
  args: unknown[];
}

/** Wire shape of a filter handler's return value. */
export interface FilterResult {
  value: unknown;
}

/**
 * Marshal an action payload to its JSON wire form.
 *
 * A null/undefined `args` becomes `[]` so the encoded envelope is
 * always `{"kind":"action","args":[]}` rather than carrying `"args":
 * null`. The guest decoder relies on the field being an array.
 */
export function marshalActionPayload(args?: readonly unknown[] | null): string {
  const payload: ActionPayload = {
    kind: PayloadKind.Action,
    args: args ? Array.from(args) : [],
  };
  return JSON.stringify(payload);
}

/**
 * Marshal a filter payload to its JSON wire form.
 *
 * Like `marshalActionPayload`, missing extras become an empty array.
 * Missing `value` becomes JSON-null so the encoded form always has the
 * field present.
 */
export function marshalFilterPayload(
  value: unknown,
  args?: readonly unknown[] | null,
): string {
  const payload: FilterPayload = {
    kind: PayloadKind.Filter,
    value: value === undefined ? null : value,
    args: args ? Array.from(args) : [],
  };
  return JSON.stringify(payload);
}

/**
 * Marshal a filter result to its JSON wire form.
 *
 * `value === undefined` is treated as JSON-null so the field is always
 * present on the wire.
 */
export function marshalFilterResult(value: unknown): string {
  const payload: FilterResult = {
    value: value === undefined ? null : value,
  };
  return JSON.stringify(payload);
}

/**
 * Parse an incoming action payload's JSON wire bytes.
 *
 * Throws a {@link CodecError} on malformed JSON, missing fields, or a
 * wrong `kind` tag. The thrown error's `status` is
 * `ResultStatus.BadPayload` so the runtime entry point can echo it back
 * to the host as a negative-length return.
 */
export function unmarshalActionPayload(raw: string): ActionPayload {
  const parsed = parseJSON(raw);
  if (!isObject(parsed)) {
    throw new CodecError(
      'action payload is not a JSON object',
      ResultStatus.BadPayload,
    );
  }
  if (parsed['kind'] !== PayloadKind.Action) {
    throw new CodecError(
      `action payload has wrong kind ${JSON.stringify(parsed['kind'])}`,
      ResultStatus.BadPayload,
    );
  }
  const args = parsed['args'];
  if (!Array.isArray(args)) {
    throw new CodecError(
      'action payload args is not an array',
      ResultStatus.BadPayload,
    );
  }
  return { kind: PayloadKind.Action, args };
}

/**
 * Parse an incoming filter payload's JSON wire bytes.
 *
 * Same error contract as {@link unmarshalActionPayload}. `value` is
 * accepted as any JSON-serialisable type, including null.
 */
export function unmarshalFilterPayload(raw: string): FilterPayload {
  const parsed = parseJSON(raw);
  if (!isObject(parsed)) {
    throw new CodecError(
      'filter payload is not a JSON object',
      ResultStatus.BadPayload,
    );
  }
  if (parsed['kind'] !== PayloadKind.Filter) {
    throw new CodecError(
      `filter payload has wrong kind ${JSON.stringify(parsed['kind'])}`,
      ResultStatus.BadPayload,
    );
  }
  if (!('value' in parsed)) {
    throw new CodecError(
      'filter payload missing value field',
      ResultStatus.BadPayload,
    );
  }
  const args = parsed['args'];
  if (!Array.isArray(args)) {
    throw new CodecError(
      'filter payload args is not an array',
      ResultStatus.BadPayload,
    );
  }
  return {
    kind: PayloadKind.Filter,
    value: parsed['value'],
    args,
  };
}

/**
 * Parse a JSON wire blob without throwing the raw `SyntaxError`. We
 * normalise to a {@link CodecError} so the runtime entry point can do
 * one `instanceof` check.
 */
function parseJSON(raw: string): unknown {
  try {
    return JSON.parse(raw);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new CodecError(`malformed JSON: ${msg}`, ResultStatus.BadPayload);
  }
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

/**
 * Typed failure carried by the codec layer.
 *
 * The runtime entry point catches this and packs `status` into the
 * low 32 bits of the i64 return. Plugin authors generally don't
 * throw `CodecError` directly — they let the codec do it for them and
 * surface their own errors via {@link PluginError} from `host.ts`.
 */
export class CodecError extends Error {
  override readonly name = 'CodecError';
  readonly status: ResultStatusCode;
  constructor(message: string, status: ResultStatusCode = ResultStatus.BadPayload) {
    super(message);
    this.status = status;
  }
}
